blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+of AWS machines, but not on GCP or Azure.
+
+Lastly, we shouldn't fail to mention GPUs. GPUs are heavily utilized by Julia
+users, but are currently not supported by `rr`. Changing this is tricky. GPUs
+often allow direct memory access between the device and the userspace program
+being recorded. If the interface to the GPU is known, this can definitely be
+mitigated, potentially at some performance cost, e.g. by double buffering. For
+GPUs with open source drivers, e.g. AMDGPU, there is active community interest
+in building such a solution. It is not an easy task, but I think it has a
+decent chance of succeeding, since open source drivers allow insight into exactly
+how the GPU works and when it will modify process memory. GPUs with proprietary
+drivers are a lot harder. For one, their interface to userspace is not necessarily
+known. Additionally, proprietary drivers tend to be a lot less well behaved
+than open source ones (the linux kernel review process tends to filter out
+the most egregious behavior). For example, proprietary drivers have been
+observed to use the userspace stack for scratch space. For those GPUs in use
+by a large number of Julia users, we hope to work with the relevant vendors to
+bring this capability to their platforms, but it may be a long road.
+
+In summary, these new capabilities are currently only supported on Intel x86_64
+chips. However, this is a bit of a chicken-and-egg problem. The hardware
+requirements are non-trivial, but not egregious. If the hardware vendors cared
+about this problem, they could definitely build hardware that enabled `rr` to work
+(and if they really cared, they could build hardware assist features to make
+`rr` much faster), but without a significant user base of such features they
+have little incentive to care. I'm hoping that by rolling out these capabilities
+widely, these incentives will start presenting themselves. Judging from the size
+of the community and discounting for those that use unsupported hardware or
+operating system, the number of users for whom this feature will be available
+like numbers in the hundreds of thousands. That isn't super large, by hardware
+vendor standards, but it isn't trivially small either. If any hardware vendor
+is interested in making this work, we'd be happy to talk to you (and as I
+mentioned, some of this is already in progress).
+
+# A word on privacy
+
+By its nature, an `rr` trace will contain any file touched by the process during
+its lifetime. In particular, this may contain things like your julia history,
+configuration files for your system, any secrets entered or read (i.e. make
+sure to use ssh-agent if your reproducer involves using SSH for authentication),
+any private code you may be using, etc. We are investigating building tooling
+to help understand what's in the trace and anonymize parts of the trace that are
+potentially sensitive, but do not have long-ranging effects in the trace. We
```

Comment:
What's "long-ranging effects"?

Reply:
Even if you don't read something from your history file, it could become unallocated garbage that later is used as uninitialized memory and causes a problem. I'll spell out this example.

Reply:
Actually, I think that's too detailed, I'll just say do not affect the trace